gadgets/trace_bind/program.bpf.c

```diff
@@ -0,0 +1,251 @@
+/* SPDX-License-Identifier: (LGPL-2.1 OR BSD-2-Clause) */
+/* Copyright (c) 2021 Hengqi Chen */
+/* Copyright (c) 2024 The Inspektor Gadget authors */
+
+#include <vmlinux.h>
+#include <bpf/bpf_helpers.h>
+#include <bpf/bpf_core_read.h>
+#include <bpf/bpf_tracing.h>
+#include <bpf/bpf_endian.h>
+
+#include <gadget/buffer.h>
+#include <gadget/macros.h>
+#include <gadget/mntns_filter.h>
+#include <gadget/types.h>
+
+enum bind_options_set : __u8 {
+	FREEBIND = 0x1,
+	TRANSPARENT = 0x2,
+	BIND_ADDRESS_NO_PORT = 0x4,
+	REUSEADDRESS = 0x8,
+	REUSEPORT = 0x10,
+};
+
+union bind_options {
+	__u8 data;
+	struct {
+		__u8 freebind : 1;
+		__u8 transparent : 1;
+		__u8 bind_address_no_port : 1;
+		__u8 reuseaddress : 1;
+		__u8 reuseport : 1;
+	} fields;
+};
+
+struct event {
+	gadget_timestamp timestamp_raw;
+	gadget_mntns_id mount_ns_id;
+	struct gadget_l4endpoint_t addr;
+	// user terminology for pid:
+	__u32 pid;
+	__u32 uid;
+	__u32 gid;
+	int ret;
+	enum bind_options_set opts_raw;
+	char comm[TASK_COMM_LEN];
+	// TODO: How to get the name of the device?
+	__u32 bound_dev_if;
+};
+
+const volatile pid_t target_pid = 0;
+const volatile bool ignore_errors = true;
+const volatile bool filter_by_port = false;
+
+GADGET_PARAM(target_pid);
+GADGET_PARAM(ignore_errors);
+// TODO: Add support for filtering by port
+//GADGET_PARAM(filter_by_port);
```

Comment:
> This would mean we would filter in user space and not in eBPF space, right?

No exactly. My idea is to have the wasm module configuring the filtering (writing the values to the eBPF map and setting the constant to true).

Reply:
Opened #3120